src/ess/bifrost/single_crystal/q_map.py

```diff
@@ -0,0 +1,247 @@
+# SPDX-License-Identifier: BSD-3-Clause
+# Copyright (c) 2025 Scipp contributors (https://github.com/scipp)
+
+"""Build a Q-map for single crystal diffraction."""
+
+from collections.abc import Callable
+from typing import Any
+
+import numpy as np
+import numpy.typing as npt
+import plopp as pp
+import scipp as sc
+from matplotlib.axes import Axes
+
+from ess.spectroscopy.types import DetectorCountsWithQ, RunType
+
+from .types import CountsWithQMapCoords, QProjection
+
+
+def project_momentum_transfer(
+    counts: DetectorCountsWithQ[RunType],
+    *,
+    q_projection: QProjection,
+) -> CountsWithQMapCoords[RunType]:
+    """Project Q onto given axes and also compute the norm of Q."""
+    transformed = counts.transform_coords(
+        ['Q_parallel', 'Q_perpendicular', 'Q'],
+        graph={
+            'Q_parallel': _make_projection_kernel(q_projection.parallel),
+            'Q_perpendicular': _make_projection_kernel(q_projection.perpendicular),
+            'Q': lambda sample_table_momentum_transfer: sc.norm(
+                sample_table_momentum_transfer
+            ),
+        },
+        keep_inputs=False,
+    )
+    if transformed.bins is not None:
+        transformed.bins.coords['a3'] = sc.bins_like(
+            transformed, transformed.coords['a3']
+        )
+        transformed = transformed.bins.concat()
+
+    return CountsWithQMapCoords[RunType](transformed)
+
+
+def _make_projection_kernel(vector: sc.Variable) -> Callable[..., sc.Variable]:
+    normal = vector / sc.norm(vector)
+
+    def projection(sample_table_momentum_transfer: sc.Variable) -> sc.Variable:
+        return sc.dot(sample_table_momentum_transfer, normal)
+
+    return projection
+
+
+def default_q_projection() -> QProjection:
+    return QProjection(
+        parallel=sc.vector(value=[0, 0, 1]), perpendicular=sc.vector(value=[1, 0, 0])
+    )
+
+
+def make_q_map(
```

Comment:
I am not clear on whether the below would be needed for ESSlivedata? If so, is there anything that can be made plotting-lib independent?

Reply:
It is just that I did not understand whether such a complicated setup code would also be needed make a regular plot.

Reply:
Or is that what you meant with
> I will add extra providers to compute the Q projections directly for live data in a separate PR. Please check that the current implementation makes sense!
?

Reply:
Yes. This setup is only complicated because the code draws the circles and annulus into the plot. Slicing itself is pretty much trivial.